Mosquitto

prince-chrismc opened the pull request:
There's a stale PR that I could love to help push forward. Not sure who's responsible for this account 🤔 

https://github.com/conan-io/conan-center-index/pull/2033#issuecomment-744244363

Croydon commented on recipes/mosquitto/all/conanfile.py:
```diff
@@ -38,7 +40,7 @@ def configure(self):
     def requirements(self):
         if self.options.with_tls:
-            self.requires("openssl/1.1.1g")
+            self.requires("openssl/1.1.1h")
```
```suggestion
            self.requires("openssl/1.1.1i")
```

Croydon commented on recipes/mosquitto/all/conanfile.py:
```diff
@@ -100,7 +108,7 @@ def package_info(self):
             self.cpp_info.components["libmosquitto"].defines.append("WITH_TLS")
         if self.settings.os == "Windows":
-            self.cpp_info.components["libmosquitto"].system_libs.append("ws2_32")
+            self.cpp_info.components["libmosquitto"].system_libs.extend(["ws2_32", "crypt32"])
```
Does it always require `crypt32` even when `self.options.with_tls` is `False`?

prince-chrismc replied:
iam doubtful but it's been too long

uilianries replied:
Are you sure about crypt32? I can't find any result when searching over the code base:

```
$ grep ws2_32 -r /home/conan/.conan/data/mosquitto/1.6.12/_/_/source/source_subfolder/ 
/home/conan/.conan/data/mosquitto/1.6.12/_/_/source/source_subfolder/src/CMakeLists.txt:	set (MOSQ_LIBS ${MOSQ_LIBS} ws2_32)
/home/conan/.conan/data/mosquitto/1.6.12/_/_/source/source_subfolder/lib/CMakeLists.txt:	set (LIBRARIES ${LIBRARIES} ws2_32)
$ grep crypt32 -r /home/conan/.conan/data/mosquitto/1.6.12/_/_/source/source_subfolder/ 
$
```

Maybe it's getting a ride from another dependency.

prince-chrismc replied:
it's most likely being pulled in from openssl

prince-chrismc replied:
now that I checked out the code, I did this because the `_patch` method forces it all the time and it was being pulled in since the default TLS is on.

I'll fix the patch and this

uilianries commented on recipes/mosquitto/all/conandata.yml:
```diff
@@ -1,5 +1,5 @@
 sources:
-  "1.6.10":
-    sha256: ffcfca01342ca8b82303bd1a9ccb0ff9b715e5711280a2fedba14e739f54f595
```
You can't remove old version already available for download.

prince-chrismc replied:
I am not sure I understand since I did (2 months ago). I see removing `1.6.10` and replacing with `1.6.12`

prince-chrismc replied:
This is against and ADD PR in CCI so this version has never been checked in

uilianries commented:
I see they released version 2.0 few weeks ago!! :rocket:

uilianries commented:
@a4z just opened the PR for Mosquitto 2.x: https://github.com/conan-io/conan-center-index/pull/3949


I don't know how much they differ from 1.x to 2.x and if we should use separated version folders for each major version.

a4z commented:
did not know that exists, but I need a useful mqtt library because the existing one is broken and the author is a disaster, hostile and incompetent so I opened also a PR

v2 has total different config options, v1 is not useful for me so I do not care

how to proceed to get mosquitto asap into cci?

uilianries commented:
@a4z Just use a version folder names `2.x` instead of `all` and we are all good. We will be able to add any other major without conflicts.

a4z commented:
I do that already (folder 2.0.3)

a4z commented:
no wrong, ok, will change

uilianries commented:
Some aspects are missing, as cmake wrapper. I'll open a PR for you fixing it.

a4z commented:
you mean, I will get an other 1000 Lines hook message .... :-) 
ok, thanks for the hint and the help Uilian!  will look litle later this evening on it

uilianries commented on recipes/mosquitto/config.yml:
```diff
@@ -1,3 +1,3 @@
 versions:
-  "1.6.10":
+  "1.6.12":
     folder: "all"
```
```suggestion
    folder: "1.x"
```

prince-chrismc commented:
I addressed the comments, can we please merge this so CI can run and make sure the rest of the configs are not broken =D

I have a few products using 1.x and could still benefit from this being merged

a4z commented:
imho, these fixes should be patches, not replace in the conan file, because with patches you see at the first glance there are changes, replace in a conan file hides that.
(haveng patches in Uilians mosquitto PR would have spared me quite some time yesterday)

prince-chrismc commented:
Patches are always applied.

`Tools.replace` is required for settings/option specific changes

a4z commented:
the proper fix should work on all platforms I guess?
because how can other consume openssl without a tool replace action ? 

and then the chance can be submitted upstream, back to the project.

in this way cci becomes an interesting place that improves the builds of project and that can contribute back, 
not motivating people to fix the same problem over and over in their personal dark corner (what is, btw, a side effect of this not very well through thought _a policy about patching_ issue) 

anyhow, I might have time the next time to give that a look and see what a universal solution in the cmake file itself could be

prince-chrismc commented:
openssl does not use cmake, it uses the standard GNU tooling and standard arguement https://github.com/openssl/openssl/search?q=fpic... which seamlessly works on GNU compilers thankfully for us

99.9% of consumers just set PIC and forget about it, but some care about the subtle difference 🤷 hence we support it in CCI.
This most likely explains why this project forces it with little attention to choice.

A Patch here would be extremely because it's needed for every single target (including the ones which are not packaged) these are subject to change over time across majors and new examples... having a patch would be more difficult in the long run than a replace to me.

To support this in cmake is probably more work than upstream will care to maintain please be kind if they decline your PR 🤗 
Just because it helps one person does not mean they are obliged to obey, they have many other things too juggle. ⚖️

a4z commented:
I was talking about finding / getting the requirement openssl libraries, not the PIC flag, and if you look at the various recipes, that seem to effect everyone who wants to link statically against openssl

a4z commented:
I am unable to make that work on Windows,  test fails with MT runtime configs due to obscure linker errors, not on other configs
https://github.com/conan-io/conan-center-index/pull/3949

since Windows is not  my platform and this linker error are too obscure for me to be able to solve them , and I do not need Windows build anyway, I give up on mosquitto. 
hope you have more luck

a4z commented:
Chris , a minor  this copies 2 time the same license file
https://github.com/bincrafters/conan-center-index/blob/5dcdb75e932ff0f282efecf24fd4c513227c2b03/recipes/mosquitto/1.x/conanfile.py#L95
One of them should be v2

one question
since when do conan recipes append cmake requirements, like  `.requires.append("openssl::openssl")` 
is this somehow new, where is it in the docs (I have the feeling I would need to read through all the docs every release, and would still not get all details)

prince-chrismc commented:
I made the same mistake https://github.com/bincrafters/conan-center-index/pull/19/commits/82cefb65df311fe297a3600d3f024cedf8230219

e`d`l vs e`p`l super hard to see 

https://github.com/bincrafters/conan-center-index/blob/5dcdb75e932ff0f282efecf24fd4c513227c2b03/recipes/mosquitto/1.x/conanfile.py#L94-L95

They changed the licensing between the majors as well it seems!

---

I had always assumed that it was because of commonents. I am far from the expert and I've truly learnt from just reviewing.

I am not familiar with the technical depth behind it.

Hopefully someone else can chime in and explain it to us =)

Croydon commented:
Done via https://github.com/conan-io/conan-center-index/pull/4288 (I think)
